File: typescript/2024/day7.ts
import { breadthFirstSearch } from 'fp-search-algorithms';
import * as R from 'ramda';

import { parseInt10 } from '../lib/fp';

const content = (await Bun.file('../inputs/2024/Day7/input.txt').text()).trim();

const parse = (line: string): [number, number[]] => {
  const [testValue, nums] = line.split(': ');
  return [parseInt10(testValue), nums.split(' ').map(parseInt10)];
};

const equations = content.split('\n').map(parse);

// console.log(equations);

type State = [acc: number, i: number];

const process =
  (ops: ((a: number, b: number) => number)[]) =>
  ([testValue, values]: [number, number[]]) => {
    const found = ([acc, idx]: State) => acc === testValue && idx === values.length - 1;
    const next = ([acc, idx]: State): State[] => {
      if (acc > testValue) return [];
      const nextIdx = idx + 1;
      if (nextIdx >= values.length) return [];
      const value = values[nextIdx];
      return ops.map<State>(fn => [fn(acc, value), nextIdx]);
    };

    const result = breadthFirstSearch(next, found, [values[0], 0]);
    return result;
  };

const r1 = R.sum(equations.filter(process([R.add, R.multiply])).map(a => a[0]));

console.log(r1);

const r2 = R.sum(equations.filter(process([R.add, R.multiply, (a, b) => parseInt10(`${a}${b}`)])).map(a => a[0]));

console.log(r2);

File: haskell/2024/Day7.hs
module Day7 where

import Algorithm.Search
import Data.List
import Data.List.Split
import Data.Maybe

parse :: String -> (Int, [Int])
parse s =
  let [a, b] = splitOn ": " s
      nums = map read $ splitOn " " b
   in (read a, nums)

process :: [Int -> Int -> Int] -> (Int, [Int]) -> Maybe [(Int, Int)]
process ops (testValue, values) = bfs next found (head values, 0)
  where
    len = length values
    found (acc, idx) = acc == testValue && idx == len - 1
    next (acc, idx)
      | acc > testValue = []
      | idx + 1 >= len = []
      | otherwise = let ni = idx + 1 in map (\f -> (f acc (values !! ni), ni)) ops

main' :: IO ()
main' = do
  content <- map parse . lines <$> readFile "../inputs/2024/Day7/input.txt"
  -- print content
  let ops = [(+), (*)]
  let r1 = sum $ map fst $ filter (isJust . process ops) content
  print r1
  let ops2 = [(+), (*), \a b -> read $ show a ++ show b]
  let r2 = sum $ map fst $ filter (isJust . process ops2) content
  print r2
